src/Compilers/Core/Portable/CommandLine/CommonCompiler.CompilerEmitStreamProvider.cs

```
@@ -40,15 +40,71 @@ public override Stream CreateStream(DiagnosticBag diagnostics)
+                        //
+                        // Note that if the file is marked "readonly" or the current user doesn't have sufficient privileges
+                        // the exception thrown is UnauthorizedAccessException, not IOException, so we won't attempt to delete the file.
+
+                        try
+                        {
+                            const int eWin32SharingViolation = unchecked((int)0x80070020);
+
+                            if (PathUtilities.IsUnixLikePlatform)
+                            {
+                                // Unix & Mac are simple: just delete the file in the directory. 
+                                // The memory mapped content remains available for the reader.
+                                PortableShim.File.Delete(_filePath);
+                            }
+                            else if (e.HResult == eWin32SharingViolation)
+                            {
+                                // On Windows File.Delete only marks the file for deletion, but doens't remove it from the directory.
+                                var newFilePath = Path.Combine(Path.GetDirectoryName(_filePath), Guid.NewGuid().ToString() + "_" + Path.GetFileName(_filePath));
+
+                                // Try to rename the existing file. This fails unless the file is open with FileShare.Delete.
+                                PortableShim.File.Move(_filePath, newFilePath);
+
+                                // hide the renamed file:
+                                PortableShim.File.SetAttributes(newFilePath, PortableShim.FileAttributes.Hidden);
+
+                                // Mark the renamed file for deletion, so that it's deleted as soon as the current reader is finished reading it
+                                PortableShim.File.Delete(newFilePath);
+                            }
+                        }
+                        catch
```

They are different - this exception is completely ignored. The other exception is converted to a diagnostic.